doc/OnlineDocs/contributed_packages/pyros.rst

```diff
@@ -0,0 +1,510 @@
+    :special-members: __init__
+
+.. autoclass:: pyomo.contrib.pyros.uncertainty_sets.UncertaintySet
+    :special-members: __init__
+
+.. autoclass:: pyomo.contrib.pyros.uncertainty_sets.DiscreteSet
+    :special-members: __init__
+
+.. autoclass:: pyomo.contrib.pyros.uncertainty_sets.IntersectionSet
+    :special-members: __init__
+
+
+PyROS Usage Example
+-----------------------------
+
+We will use an example to illustrate the usage of PyROS. The problem we will use is called *hydro* and comes from the GAMS example problem database in `The GAMS Model Library <https://www.gams.com/latest/gamslib_ml/libhtml/>`_. The model was converted to Pyomo format via the `GAMS Convert tool <https://www.gams.com/latest/docs/S_CONVERT.html>`_.
+
+This model is a QCQP with 31 variables. Of these variables, 13 represent degrees of freedom, with the additional 18 being state variables.
+The model features 6 linear inequality constraints, 6 linear equality constraints, 6 non-linear (quadratic) equalities, and a quadratic objective.
+We have augmented this model by converting one objective coefficient, two constraint coefficients, and one constraint right-hand side into Param objects so that they can be considered uncertain later on.
+
+.. note::
+    Per our analysis, the *hydro* problem satisfies the requirement that each value of :math:`\left(x, z, q \right)` maps to a unique value of :math:`y`, which indicates a proper partition of variables between (first- or second-stage) degrees of freedom and state variables.
+
+Step 0: Import the PyROS Module
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In anticipation of using the PyROS solver:
+
+.. code::
+
+  >>> # === Required import ===
+  >>> import pyomo.pyros as pyros
+
+  >>> # === Instantiate the PyROS solver object ===
+  >>> pyros_solver = SolverFactory("pyros")
+
+Step 1: Define the Deterministic Problem
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The deterministic Pyomo model for *hydro* is shown below.
+
+.. note::
+    Primitive data (Python literals) that have been hard-coded within a deterministic model cannot be later considered uncertain, unless they are first converted to ``Param`` objects within the ``ConcreteModel`` object.
+    Furthermore, any ``Param`` object that is to be later considered uncertain must have the property ``mutable=True``.
+
+.. note::
+    In case modifying the ``mutable`` property inside the deterministic model object itself is not straight-forward in your context,
+    you may consider adding the following statement **after** ``import pyomo.environ as pyo`` but **before** defining the model object:
+    ``pyo.Param.DefaultMutable = True``. Note how this sets the default ``mutable`` property in all ``Param`` objects in the ensuing model instance to ``True``;
+    consequently, this solution will not work with ``Param`` objects for which the ``mutable=False`` property was explicitly enabled inside the model object.
+
+.. code::
```

Comment:
Will the documentation tests have access to BARON with CPLEX? This is the solver I used in my own testing, I'm not sure how these will behave without it.

Reply:
It depends on the platform. On GHA, it is unlicensed BARON. On Jenkins we have licensed BARON. You can set the tests up to be skipped if the solver you want is not available / licensed. Take a look at the (I think) GDPopt or Mindtpy docs for examples.

Reply:
I see that in `conf.py` the variables determining the availability of solvers (like `baron_available`) are currently commented out. Would that need to be brought back in to check availability and license availability?

Reply:
Those are actually not commented out: `doctest_global_setup` is a string that is `eval`'ed by doctest before it evaluates the `:skipif:` directives / the code snippets.

Reply:
Makes sense to me. Since #2011 was merged, you might be able to get away with changing `conf.py` to `baron = pyomo.opt.SolverFactory('baron')`; then the `skipif:`'s can be `baron.available()` and/or `baron.license_is_valid()`